FEAT: Support context adjustment for udfs for pandas backend

LeeTZ opened the pull request:
Overview
=====
This PR add context adjustment for udfs for pandas backend.
Previously, window context adjustment is tested with built-in aggregations.  Context adjustment will fail when user put udf in window aggregations, for example:
```
@udf.reduction(['double'], 'double')
def my_mean(series):
    return series.mean()

window = ibis.trailing_window(ibis.interval(days=3), order_by='time')
expr = time_table.mutate(v1=my_mean(time_table['value']).over(window))
result = expr.execute(timecontext=context)
```
`result` will look like:
```
                     time   id  value    v1
0 2017-01-05 01:02:03.234  4.0    4.4  1.00
1 2017-01-06 01:02:03.234  5.0    5.5  1.50
2 2017-01-07 01:02:03.234  6.0    6.6  1.65
3 2017-01-08 01:02:03.234  7.0    7.7  2.75
4 2017-01-09 01:02:03.234  8.0    8.8  3.85
5                     NaT  NaN    NaN  4.95
6                     NaT  NaN    NaN  6.05
7                     NaT  NaN    NaN  7.15
```
Extra rows are added since the result Series of udf is not trimmed by timecontext.
We did implement rules to trim result Series in window execution. However it is based on a fact that the Series contains a time col, so we could extract time info and trim according to time. The result Series of udf doesn't include a time index in current implementation. This PR adds 'time' col (if present) as index for the result of udf execution. To match the result Series of built-in aggregations.

How is this tested
=====
Test is added in `ibis/backends/pandas/tests/execution/test_timecontext.py`  for udf use cases, including unit test for the core method`construct_time_context_aware_series`, and integration test for udfs on different windows (when there is no groupby nor orderby, both groupby and orderby , only orderby etc.)

jreback commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,40 @@ def canonicalize_context(
+    pd.Series
+    """
+    if TIME_COL == frame.index.name:
+        time_index = frame.index
```
are all of these cases hit in tests

LeeTZ replied:
Yes this function is unit tested in `test_timecontext` for different frame / series settings. (e.g. frame is time indexed or not, series contains time index already or not.) see `test_construct_time_context_aware_series`

jreback commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,40 @@ def canonicalize_context(
     return begin, end
 
 
+def construct_multi_index_series(
```
can we have a more informative name for this. yes its constructing a MI series, but its doing it for time_context.
maybe

construct_time_context_aware_series

icexelloss replied:
+1

LeeTZ replied:
Good idea, I changed the name to be `construct_time_context_aware_series`

jreback reviewed (changes requested):
can you add a release note as well

icexelloss commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,40 @@ def canonicalize_context(
+def construct_multi_index_series(
+    series: pd.Series, frame: pd.DataFrame
+) -> pd.Series:
+    """ Construct a pd.MultiIndex of IntIndex and 'time'
```
Why IntIndex?

icexelloss replied:
Is the IntIndex here presenting the groupby key? What about the case that groupby key is not integer?

LeeTZ replied:
I rephrase this a little bit, it meant to add 'time' into the original index, whatever type it is.

icexelloss commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,40 @@ def canonicalize_context(
+    series: pd.Series
+    frame: pd.DataFrame
+
+    Returns
```
Can you add examples of shape of the index in various cases?

LeeTZ replied:
Sure, different shapes of the index is unit tested in `test_timecontext.py`, but I also add actual examples in docstring for easier understading

icexelloss commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,40 @@ def canonicalize_context(
+    elif TIME_COL in frame:
+        time_index = frame.set_index(TIME_COL).index
+    else:
+        raise com.IbisError(f'"time" column not present in DataFrame {frame}')
```
When do we raise this error? We do not always require a time column, right?

LeeTZ replied:
You are right, this error raises when 'time' column is not present. Note that any reference to this function there is a if statement, only when there is `timecontext` defined will we call this `construct_time_context_aware_series`.

icexelloss replied:
Can we assert when entering the function that time column exists?

LeeTZ replied:
Yes, it's just the same as `assert TIME_COL in frame or TIME_COL == frame.index.name` when entering the function, it's just because I want to branch for these two cases again to extract time_index so I put this as a raise error in else branch.

icexelloss replied:
Oh ok

icexelloss commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,40 @@ def canonicalize_context(
+    if TIME_COL == frame.index.name:
+        time_index = frame.index
+    elif TIME_COL in frame:
+        time_index = frame.set_index(TIME_COL).index
```
Probably question for Jeff: Is this the best way to construct an index from existing column?

LeeTZ replied:
@jreback ^

jreback replied:
no ,do this
`pd.Index(frame[TIME_COL])`

icexelloss commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,40 @@ def canonicalize_context(
+
+    Parameters
+    ----------
+    series: pd.Series
```
Can you describe what is the series and what is the frame here?

LeeTZ replied:
Same as comments above, added description, as well as examples.

icexelloss commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,40 @@ def canonicalize_context(
     return begin, end
 
 
+def construct_multi_index_series(
```
It also feels to me what the function is an implementation of the pandas backend and therefore, probably don't belong in ibis/expr/timecontext.py module. Is there a place you can put in the pandas module?

LeeTZ replied:
I thought about this. This could be a util function in general, also putting this in pandas/execution/timecontext.py seems to cause circular dependency. I could revisit.

icexelloss replied:
Why would it have a circular dependency?

icexelloss commented on ibis/backends/pandas/execution/window.py:
```diff
@@ -338,6 +341,8 @@ def execute_window_op(
     series = post_process(result, data, ordering_keys, grouping_keys)
 
+    if timecontext:
+        series = construct_multi_index_series(series, data)
```
Are the series and data matching in row ordering here?

icexelloss replied:
i.e. I think the series is ordered in group_by ordering, and the data is ordered in original ordering (probably ordered by time).

LeeTZ replied:
It's a good point. I made some modifications here. Time index should be injected to series prior to groupby. I add logics in `aggcontext.py` for this, row order is well tracked in `agg` phase.

icexelloss reviewed (changes requested):
Left some comments.

LeeTZ commented:
Update with a change in `window.py` and `aggcontext.py` to address row ordering for the series and parent DataFrame, when groupby is defined.
Also added more tests, including unit test for `onstruct_time_context_aware_series`, and integration test for udfs when there is no `groupby` nor `orderby`, both `groupby` and `orderby` , and only `orderby`
A relase note is also added.

icexelloss commented on ibis/backends/pandas/execution/window.py:
```diff
@@ -32,19 +33,27 @@
     if isinstance(result, pd.Series):
         # `result` is a Series when an analytic operation is being
         # applied over the window, since analytic operations are N->N
+        if timecontext:
```
Why do we need this here if this is already handled inside aggcontext?

LeeTZ replied:
We need this since it's not fully handled in aggcontext.

In `aggcontext.py`, for class `Window`, method `agg`, there are 3 different execution paths:
1. built-in aggregation, this is handled by `window_agg_built_in` in line 649, it returns a multiindex series and context adjustment(trimming) works well.
2. udfs, this is handled by `window_agg_udf` in line 634, and it is the main change we made. 
3. However there is another shortcut, in line 547, for cases where there is not groupby nor orderby. It is just calling pandas methods, the result could be either a series(analytic operations), or just a number(reduction operations). This `post_process_empty` here is meant to deal with this case, to return the series, or extend the result into a series. In order for this series to be trimmed, we need to add time index here.

icexelloss replied:
I see. Can you add comments here to explain?

LeeTZ replied:
Sure, I added an inline comment for this

icexelloss commented on ibis/backends/pandas/execution/window.py:
```diff
@@ -178,12 +193,14 @@ def trim_with_timecontext(data, timecontext: Optional[TimeContext]):
-    # indexes are retturned.
-    if TIME_COL not in df:
+    # indexes are returned.
+    if TIME_COL not in df or not is_datetime64_dtype(df[TIME_COL]):
```
Same as above, why do we need this here?

LeeTZ replied:
Not necessarily related with udf context adjustment, I saw typos in the comment, and just add a further check here to prevent the following operations to throw an exception, if `time` is of an unexpected, non-DateTime type

icexelloss replied:
> if time is of an unexpected, non-DateTime type

Can you point me to the test that covers this?

LeeTZ replied:
Yes, see a unit test for this function [test_trim_with_context](https://github.com/ibis-project/ibis/pull/2646/files#diff-4ca17da469238b703a381003848b264dffdf3474fd0d3070b118ddec5e4dc16fR138)

icexelloss replied:
I see. Is there a user/API level test that could hit this?

icexelloss replied:
Also it's not clear why sometimes you are checking the types the TIME_COL and sometimes not. So it's bit confusing when we should check the type and when should we just check existance.

LeeTZ replied:
Hmm, no user tests yet. For checking, I think to be strict, `TIME_COL` must be present and the dtype should be of `datetime64`, only then we could safely apply time based operations like these.  There are some cases indeed it's just checking the existence like `if TIME_COL` or `if timecontext`, I think it's because the following actions are not directly doing operations on timestamp type e.g. trimming by comparing the value of the column with a `pd.Timestamp`.

icexelloss replied:
ok

icexelloss commented on ibis/backends/pandas/tests/execution/conftest.py:
```diff
@@ -144,7 +144,7 @@ def time_df3():
                 ).values
             ),
-            'id': list(range(1, 9)),
+            'id': list(range(1, 5)) * 2,
```
Why changing this?

LeeTZ replied:
This is meant to extend the test cases to show that `group_by` parameter in window could actually changed the row ordering.  When `id` are all distinct, `groupby` produces a result Dataframe where the row ordering is the same with the case where there is no `groupby`, but with this change, when we group by `id` in tests, the udf result is not guranteed to be ordered by `time`.  A new test is added to show in this case we still trim data with time context right.

LeeTZ replied:
`test_context_adjustment_window_udf_groupby` in `test_timecontext.py`

icexelloss replied:
OK

icexelloss commented on ibis/backends/pandas/execution/window.py:
```diff
@@ -53,13 +62,18 @@ def _post_process_group_by(
     parent: pd.DataFrame,
     order_by: List[str],
     group_by: List[str],
+    **kwargs,
```
Should this be timecontext instead of kwargs?

LeeTZ replied:
It could be, but `timecontext` is currently not used in this function so I didn't make it as an explicit param

icexelloss replied:
I think it's a good pattern to specify all param explicitly even some of them are not used. kwargs brings flexibility which I don't think you need it here.

icexelloss replied:
(Explicitness over implicitness)

jreback replied:
agree with @icexelloss here, we don't want kwargs at all.

LeeTZ replied:
thx, I removed `kwargs`

icexelloss commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,88 @@ def canonicalize_context(
+    else:
+        raise com.IbisError(f'"time" column not present in DataFrame {frame}')
+    if TIME_COL not in series.index.names:
+        series.index = pd.MultiIndex.from_arrays(
```
You probably want Jeff to take a look at this code. I don't know if this is the best way.

icexelloss replied:
Does the ordering of indexes matter? e.g., key, time vs time, key

LeeTZ replied:
Yeah, this is similar to implementation in [window_agg_built_in](https://github.com/ibis-project/ibis/blob/08261f53f7685f44754f52741e025473aecef762/ibis/backends/pandas/aggcontext.py#L430)

LeeTZ replied:
In terms of trimming, no, when we trim by timecontext MultiIndex is flatten, in [trim_with_timecontext](https://github.com/ibis-project/ibis/blob/08261f53f7685f44754f52741e025473aecef762/ibis/backends/pandas/execution/window.py#L195)

icexelloss replied:
Gotcha

icexelloss commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,88 @@ def canonicalize_context(
+        1 2017-01-03 01:02:03.234   2    2.2
+        2 2017-01-04 01:02:03.234   3    3.3
+
+        For a series of:
```
Can you add some index here for better readability ?

1. For a series:
  ...
   The result series will be

2. For a series:
    ...
    The result will be
    ...

LeeTZ replied:
Good point, combining with Jeff's comment I update this to be runnable code / ipython output

jreback commented on ibis/backends/pandas/tests/execution/test_timecontext.py:
```diff
@@ -222,3 +263,121 @@ def test_context_adjustment_window_groupby_id(time_table, time_df3):
     tm.assert_series_equal(result, expected)
+
+
+def test_construct_time_context_aware_series(time_df3):
```
can you parameterize this? or too crazy

LeeTZ replied:
I tried to parameterize this but found `expected` quite different from each other, also want to highlight differences of the existence of groupby & orderby so split these

jreback replied:
ok sure

if you can put a blank line fore comments (e.g. this is basically a new case). and if needed break this into separate tests to separate the cases as this is pretty long

LeeTZ replied:
+1

jreback commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,88 @@ def canonicalize_context(
+    Examples:
+        Assume frame is:
+                             time  id  value
+        0 2017-01-02 01:02:03.234   1    1.1
```
can you actually programatically construct this, *then* show it.

also use
```
Examples
-----------
```

jreback replied:
IOW do this in ipython and basically copy it here (or alternatly do it in a python terminal and use `>>>`

jreback replied:
idea is that this code can be copy pasted to a terminal for execution directly; so include imports too.

LeeTZ replied:
Good point thx

jreback commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +127,88 @@ def canonicalize_context(
+        Name: value, dtype: float64
+
+        The result series will be:
+                             time
```
move all examples *after* Parameters

LeeTZ commented on ibis/backends/pandas/tests/execution/test_timecontext.py:
```diff
@@ -222,3 +263,121 @@ def test_context_adjustment_window_groupby_id(time_table, time_df3):
+    return series.mean()
+
+
+def test_context_adjustment_window_udf_nogroupby_noorderby(
```
+1

icexelloss commented on ibis/backends/pandas/tests/execution/test_timecontext.py:
```diff
@@ -132,6 +134,44 @@ def test_context_adjustment_window(
 
+def test_trim_with_timecontext(time_df3):
+    """ Unit test `trim_with_timecontext` in Window execution
+    """
```
Move this to the same line? (or is it too long)

LeeTZ replied:
Seems `black` formatted this?

icexelloss replied:
kk

icexelloss commented on ibis/backends/pandas/aggcontext.py:
```diff
@@ -620,13 +621,23 @@ def agg(
+                # timecontext. In order to do so, 'time' must be added
+                # as an index to the Series, if present. Here We extract
+                # time column from the parent Dataframe `frame`.
+                if TIME_COL in frame or TIME_COL in frame.index.names:
```
When would TIME_COL be in frame.index.names? Shouldn't time always be a column?

LeeTZ replied:
There are cases when `TIME_COL` used as the index for df, aren't there?

LeeTZ replied:
I see your point of removing this, ibis should not set TIME_COL to be index

LeeTZ commented:
Move tests for timecontext for udfs to be under `ibis/backends/test` to test for all backends that support this feature (pandas & pyspark in this case). Tests are parameterized and use `alltypes` as the data source. Note that in order to make this work, I also move `TIME_COL` from a constant in `timecontext.py` to be an ibis config that is configurable on test runs.

icexelloss commented on ibis/backends/pandas/execution/generic.py:
```diff
@@ -920,6 +920,7 @@ def execute_database_table_client(
     df = client.dictionary[op.name]
     if timecontext:
         begin, end = timecontext
+        TIME_COL = get_time_col()
```
This looks funky, why are we setting a constant here?

icexelloss replied:
ok yeah, upper case is confusing because it looks like you are reassigning value to a constant.

jreback commented on ibis/backends/pandas/execution/window.py:
```diff
@@ -178,12 +204,15 @@ def trim_with_timecontext(data, timecontext: Optional[TimeContext]):
-    # indexes are retturned.
-    if TIME_COL not in df:
+    # indexes are returned.
+    TIME_COL = get_time_col()
```
same comment here (as you describe above)

jreback commented on ibis/backends/pandas/tests/execution/test_timecontext.py:
```diff
@@ -132,6 +134,43 @@ def test_context_adjustment_window(
+    """ Unit test `trim_with_timecontext` in Window execution"""
+    df = time_df3.copy()
+    context = pd.Timestamp('20170105'), pd.Timestamp('20170111')
+    # trim_with_timecontext takes a MultiIndex Series as input
```
can you put a blank line before a comment

jreback commented on ibis/backends/tests/test_timecontext.py:
```diff
@@ -0,0 +1,100 @@
+        .rolling('3d', closed='both')[TARGET_COL]
+        .mean()
+    ).reset_index()
+    # Result is a MultiIndexed Series
```
blank line

LeeTZ replied:
Yeah I have added blank lines between these test cases

jreback commented on ibis/backends/tests/test_timecontext.py:
```diff
@@ -0,0 +1,100 @@
+    return [exp_raw_win, exp_orderby, exp_groupby_orderby]
+
+
+@pytest.mark.only_on_backends(['pandas', 'pyspark'])
```
are these new for pyspark? i thought you had tests already ? dask? (just asking absolutely could be separate / followup)

LeeTZ replied:
Yeah these were tests that I first added to pandas backend, while we found that pyspark backend doesn't have tests to cover  such udf use cases I extend this for pyspark backend as well. I could extend this to dask after time context is implemented for dask backend, sure.

jreback commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +132,97 @@ def canonicalize_context(
+    pd.Series
+
+    Examples
+    -----------
```
underline should be the same length

LeeTZ replied:
+1

jreback commented on ibis/expr/timecontext.py:
```diff
@@ -127,6 +132,97 @@ def canonicalize_context(
+    Name: value, dtype: float64
+    The result is unchanged for a series already has 'time' as its index.
+    """
+    TIME_COL = get_time_col()
```
same comment as above (make this lowercase)

icexelloss reviewed (approved):
LGTM

icexelloss commented on ibis/__init__.py:
```diff
@@ -33,6 +33,13 @@
 )
 ibis.config.register_option('default_backend', None)
+ibis.config.register_option(
+    'time_col',
```
Can be a follow up. We should properly namespace this to be 
`context_adjustment.time_col`

jreback commented:
thanks @LeeTZ very nice!
